File: go/pkg/speech/tts.go
package speech

import (
	"context"
	"fmt"
	"io"
	"log/slog"

	"github.com/haivivi/giztoy/pkg/audio/pcm"
	"github.com/haivivi/giztoy/pkg/trie"
)

// TTSMux is the default multiplexer for TTS synthesizers.
var TTSMux = NewTTSMux()

// HandleTTS registers a synthesizer for the given pattern with the default mux.
func HandleTTS(pattern string, synthesizer Synthesizer) error {
	return TTSMux.Handle(pattern, synthesizer)
}

// HandleTTSFunc registers a synthesizer function for the given pattern with the default mux.
func HandleTTSFunc(pattern string, f SynthesizeFunc) error {
	return TTSMux.HandleFunc(pattern, f)
}

// Synthesize performs text-to-speech synthesis using the default multiplexer.
func Synthesize(ctx context.Context, name string, textStream io.Reader, format pcm.Format) (Speech, error) {
	return TTSMux.Synthesize(ctx, name, textStream, format)
}

// Revoice synthesizes new speech audio by passing the decoded audio of the
// input speech through the specified TTS synthesizer via the default
// multiplexer.
func Revoice(ctx context.Context, name string, speech Speech, format pcm.Format) (Speech, error) {
	pr, pw := io.Pipe()
	go func() {
		if _, err := CopySpeech(pcm.Discard, pw, speech); err != nil {
			pw.CloseWithError(err)
		} else {
			pw.Close()
		}
	}()
	return Synthesize(ctx, name, pr, format)
}

// SentenceIterator is an interface for iterating over sentences.
type SentenceIterator interface {
	// Next returns the next sentence.
	Next() (string, error)
	// Close closes the iterator.
	Close()
}

// SentenceSegmenter is an interface for segmenting text into sentences.
type SentenceSegmenter interface {
	// Segment segments the text from the reader into sentences.
	Segment(io.Reader) (SentenceIterator, error)
}

// AudioDecoder is an interface for decoding audio data.
type AudioDecoder interface {
	// Decode decodes the audio data from the reader into the given PCM format.
	Decode(r io.Reader, format pcm.Format) (io.Reader, error)
}

// TextSynthesizer is an interface for synthesizing text into audio data.
type TextSynthesizer interface {
	// SynthesizeText synthesizes the text into audio data and writes it to the writer.
	SynthesizeText(ctx context.Context, text string, w io.Writer) error
}

// Synthesizer is an interface for a text-to-speech synthesizer.
type Synthesizer interface {
	// Synthesize synthesizes the text from the reader into speech.
	Synthesize(ctx context.Context, name string, textStream io.Reader, format pcm.Format) (Speech, error)
}

// SynthesizeFunc is a function that implements the Synthesizer interface.
type SynthesizeFunc func(ctx context.Context, name string, textStream io.Reader, format pcm.Format) (Speech, error)

// Synthesize implements the Synthesizer interface.
func (f SynthesizeFunc) Synthesize(ctx context.Context, name string, textStream io.Reader, format pcm.Format) (Speech, error) {
	return f(ctx, name, textStream, format)
}

// TTS is a multiplexer for TTS synthesizers.
type TTS struct {
	mux *trie.Trie[Synthesizer]
}

var _ Synthesizer = (*TTS)(nil)

// NewTTSMux creates a new TTS multiplexer.
func NewTTSMux() *TTS {
	return &TTS{
		mux: trie.New[Synthesizer](),
	}
}

// Handle registers a synthesizer for the given name.
func (r *TTS) Handle(name string, synthesizer Synthesizer) error {
	return r.mux.Set(name, func(ptr *Synthesizer, existed bool) error {
		*ptr = synthesizer
		if existed {
			slog.Warn("tts: synthesizer already registered", "name", name)
		}
		return nil
	})
}

// HandleFunc registers a synthesizer function for the given name.
func (r *TTS) HandleFunc(name string, f SynthesizeFunc) error {
	return r.Handle(name, f)
}

// Synthesize synthesizes speech for the given name.
func (r *TTS) Synthesize(ctx context.Context, name string, textStream io.Reader, format pcm.Format) (Speech, error) {
	syn, ok := r.mux.GetValue(name)
	if !ok || syn == nil {
		return nil, fmt.Errorf("tts: synthesizer not found for %s", name)
	}
	return syn.Synthesize(ctx, name, textStream, format)
}

File: rust/speech/src/tts.rs
//! Text-to-speech synthesis.

use crate::{Speech, SpeechError};
use async_trait::async_trait;
use giztoy_audio::pcm::Format;
use giztoy_trie::Trie;
use std::sync::Arc;
use tokio::io::AsyncRead;
use tokio::sync::RwLock;
use tracing::warn;

/// Error type for TTS operations.
#[derive(Debug, thiserror::Error)]
pub enum TTSError {
    #[error("synthesizer not found: {0}")]
    NotFound(String),
    #[error("synthesis failed: {0}")]
    SynthesisFailed(String),
    #[error("speech error: {0}")]
    Speech(#[from] SpeechError),
    #[error("pattern error: {0}")]
    Pattern(String),
    #[error("other error: {0}")]
    Other(String),
}

/// Interface for a text-to-speech synthesizer.
#[async_trait]
pub trait Synthesizer: Send + Sync {
    /// Synthesizes the text from the reader into speech.
    async fn synthesize(
        &self,
        name: &str,
        text_stream: Box<dyn AsyncRead + Send + Unpin>,
        format: Format,
    ) -> Result<Box<dyn Speech>, TTSError>;
}

/// Function type that implements the Synthesizer interface.
pub type SynthesizeFunc = Box<
    dyn Fn(
            String,
            Box<dyn AsyncRead + Send + Unpin>,
            Format,
        ) -> std::pin::Pin<Box<dyn std::future::Future<Output = Result<Box<dyn Speech>, TTSError>> + Send>>
        + Send
        + Sync,
>;

/// A multiplexer for TTS synthesizers.
pub struct TTS {
    mux: Arc<RwLock<Trie<Arc<dyn Synthesizer>>>>,
}

impl Default for TTS {
    fn default() -> Self {
        Self::new()
    }
}

impl TTS {
    /// Creates a new TTS multiplexer.
    pub fn new() -> Self {
        Self {
            mux: Arc::new(RwLock::new(Trie::new())),
        }
    }

    /// Registers a synthesizer for the given name pattern.
    pub async fn handle(&self, name: &str, synthesizer: Arc<dyn Synthesizer>) -> Result<(), TTSError> {
        let mut mux = self.mux.write().await;
        mux.set(name, |existing| {
            if existing.is_some() {
                warn!(name = %name, "tts: synthesizer already registered");
            }
            Ok::<_, giztoy_trie::InvalidPatternError>(synthesizer)
        })
        .map_err(|e| TTSError::Pattern(e.to_string()))
    }

    /// Synthesizes speech for the given name.
    pub async fn synthesize(
        &self,
        name: &str,
        text_stream: Box<dyn AsyncRead + Send + Unpin>,
        format: Format,
    ) -> Result<Box<dyn Speech>, TTSError> {
        let mux = self.mux.read().await;
        let syn = mux.get(name).ok_or_else(|| TTSError::NotFound(name.to_string()))?;
        syn.synthesize(name, text_stream, format).await
    }
}

#[async_trait]
impl Synthesizer for TTS {
    async fn synthesize(
        &self,
        name: &str,
        text_stream: Box<dyn AsyncRead + Send + Unpin>,
        format: Format,
    ) -> Result<Box<dyn Speech>, TTSError> {
        TTS::synthesize(self, name, text_stream, format).await
    }
}

#[cfg(test)]
mod tts_tests {
    use super::*;

    #[test]
    fn test_tts_error_display() {
        let err = TTSError::NotFound("voice/en".to_string());
        assert!(err.to_string().contains("voice/en"));

        let err = TTSError::SynthesisFailed("timeout".to_string());
        assert!(err.to_string().contains("timeout"));
    }

    #[tokio::test]
    async fn test_tts_not_found() {
        let tts = TTS::new();
        let result = tts
            .synthesize(
                "nonexistent",
                Box::new(tokio::io::empty()),
                Format::L16Mono16K,
            )
            .await;
        assert!(matches!(result, Err(TTSError::NotFound(_))));
    }
}
